src/MIDebugEngine/Engine.Impl/ExceptionManager.cs

```diff
@@ -480,17 +525,30 @@ private async Task UpdateCatagory(Guid categoryId, ExceptionCategorySettings cat
                     }
                 }
 
-                IEnumerable<ulong> breakpointIds = await _commandFactory.SetExceptionBreakpoints(categoryId, exceptionNames, grouping.Key);
-
-                int count = exceptionNames.Zip(breakpointIds, (exceptionName, breakpointId) =>
+                bool isBreakThrown = grouping.Key.HasFlag(ExceptionBreakpointStates.BreakThrown);
+                if (!IsCppExceptionsCategoryChecked(updates) && isBreakThrown)
+                {
+                    // IEnumerable<ulong> breakpointIds = await _commandFactory.SetExceptionBreakpoints(categoryId, exceptionNames, grouping.Key);
+                    // this approach is very messy, but we need to update categorySettings.CurrentRules
+                    foreach (string exceptionName in exceptionNames)
+                    {
+                        IEnumerable<ulong> breakpointIds = await _commandFactory.SetExceptionBreakpoints(categoryId, new string[] { exceptionName }, grouping.Key);
+                        categorySettings.CurrentRules.Add(exceptionName, breakpointIds.First());
+                    }
```

Comment:
Rather than sending the exceptions one at a time, we should really use the old code here.

Reply:
I understand this. That said, I did a `foreach` here because `CurrentRules` keeps track of the `exceptionName` and `breakpointId` pairs. Unless there is a better way of tracking this (perhaps with a lambda expression?)...

Notice that {exceptionName, breakpointId} is added to `CurrentRules` after `SetExceptionsBreakpoints()` is called. And notice {exceptionName, breakpointId} is removed from `CurrentRules` after `RemoveExceptionBreakpoint()` is called.

Reply:
That is why the old code was using `.Zip` to zip together the two enumerables:

```
                IEnumerable<ulong> breakpointIds = await _commandFactory.SetExceptionBreakpoints(categoryId, exceptionNames, grouping.Key);

                int count = exceptionNames.Zip(breakpointIds, (exceptionName, breakpointId) =>
                {
                    categorySettings.CurrentRules[exceptionName] = breakpointId;
                    return 1;
                }).Sum();

#if DEBUG
                Debug.Assert(count == exceptionNames.Count());
#endif
```